pkg/interface/src/views/apps/chat/components/lib/chat-window.tsx

```diff
@@ -126,68 +142,104 @@ export class ChatWindow extends Component {
     }
   }
 
-  render() {
-    const { props, state } = this;
-    const sliceLength = Math.min(
-      state.numPages * PAGE_SIZE,
-      props.messages.length + props.pendingMessages.length
-    );
-    const messages =
-      props.pendingMessages
-        .concat(props.messages)
-        .slice(0, sliceLength);
 
+  render() {
+    const { props: {
+      envelopes,
+      stationPendingMessages,
+      unreadCount,
+      unreadMsg,
+      isChatLoading,
+      isChatUnsynced,
+      api,
+      ship,
+      station,
+      association,
+      group,
+      contacts,
+      mailboxSize
+    }, state } = this;
+    
+    const messages: Envelope[] = [];
+
+    // Reindex messages based on their number.
+    // We want the virtual list to know the total
+    // size of messages and render based on that
+    envelopes
+      .concat(stationPendingMessages)
+      .forEach((message) => {
+        messages[message.number] = message;
+      });
+
+    const loadedMessageCount = Object.entries(messages).length;
+
+    // Virtuoso tries to scroll offscreen if the latest message
+    // can't reach the bottom, so just don't set the initial
+    // index if fewer than 20 messages exist in the chat
+    const initialIndex =
+      mailboxSize - 1 < INITIAL_LOAD
+      ? undefined
+      : unreadCount // otherwise if there are unread messages
+        ? mailboxSize - unreadCount - 1 // put the one right before at the top
+        : mailboxSize - 1; // Just load it at the end
+    
+    const overscan = 100;
     return (
       <Fragment>
         <UnreadNotice
-          unreadCount={props.unreadCount}
-          unreadMsg={props.unreadMsg}
+          unreadCount={unreadCount}
+          unreadMsg={unreadMsg}
           dismissUnread={this.dismissUnread} />
-        <ChatScrollContainer
-          ref={this.scrollReference}
-          scrollIsAtBottom={this.scrollIsAtBottom}
-          scrollIsAtTop={this.scrollIsAtTop}>
-          <BacklogElement isChatLoading={props.isChatLoading} />
-          <ResubscribeElement
-            api={props.api}
-            host={props.ship}
-            station={props.station}
-            isChatUnsynced={props.isChatUnsynced}
-          />
-          { messages.map((msg, i) => (
-              <ChatMessage
-                key={msg.uid}
-                unreadRef={this.unreadReference}
-                isLastUnread={
-                  props.unreadCount > 0 &&
-                  i === props.unreadCount - 1 &&
-                  state.numPages !== 1
-                }
-                msg={msg}
-                previousMsg={messages[i - 1]}
-                nextMsg={messages[i + 1]}
-                association={props.association}
-                group={props.group}
-                contacts={props.contacts} />
-            ))
-          }
-        </ChatScrollContainer>
+        <BacklogElement isChatLoading={isChatLoading} />
+        <ResubscribeElement {...{ api, host: ship, station, isChatUnsynced}} />
```

Comment:
woah I didn't realise you could do this

Reply:
Yeah, it's the same feature as 
const { props } = this